common/workunit/wujobq.cpp

```diff
@@ -1160,17 +1186,33 @@ class CJobQueue: public CJobQueueBase, implements IJobQueue
         return best;
     }
 
-    void setWaiting(unsigned numqueues,sQueueData **queues, bool set)
+    void setWaiting(unsigned numqueues,sQueueData **queues, unsigned __int64 clientPrio, bool set)
     {
         for (unsigned i=0; i<numqueues; i++) {
             IPropertyTree *croot = queryClientRootSession(*queues[i]);
             croot->setPropInt64("@waiting",croot->getPropInt64("@waiting",0)+(set?1:-1));
+            //If a non-zero client priority has been specified, add (or remove) it from the list of priorities
+            if (clientPrio)
+            {
+                if (set)
+                    croot->setPropInt64("@priority", clientPrio);
+                else
+                    croot->removeProp("@priority");
+            }
         }
     }
 
 // 'simple' queuing
-    IJobQueueItem *dodequeue(int minprio,unsigned timeout=INFINITE, bool useprev=false, bool *timedout=NULL)
+    IJobQueueItem *dodequeue(int minprio, __uint64 clientPrio, unsigned timeout, bool useprev, bool * timedout)
     {
+        //If more than one thread is waiting on the queue, then the queue code does not work correctly
+        //It is undefined which thread the semaphore signal will wake up.
+        //E.g. there is one thread with a minimum priority of 0, and another with a minimum of 100, and an item of
+        //priority 50 is queued.  If the minimum priority of 100 is woken twice nothing will be dequeued.
+        //Similar problems occur when the clientPriority is mixed.
+        if (isProcessingDequeue.exchange(true))
```

Comment:
Should you sleep a tiny amount and try again instead of throwing ?
What is the implication if we throw ?

Reply:
ok, I see comment -
no component has more than one thread listening to a single job queue object.